Sources/Datadog/RUM/FrustrationSignals/RageTapManager.swift

```diff
@@ -0,0 +1,154 @@
+    
+    
+    func canAcceptTapIntoChain(chain: RageTapChain, tapLocation: CGPoint) -> Bool{
+        if chain.actionCommands.count == 0 {
+            return true
+        }
+        
+        let sqDistanceX = (chain.lastTapPosition.x - tapLocation.x) * (chain.lastTapPosition.x - tapLocation.x)
+        let sqDistanceY = (chain.lastTapPosition.y - tapLocation.y) * (chain.lastTapPosition.y - tapLocation.y)
+
+        if (sqDistanceX + sqDistanceY) <= (RageTapManager.Constants.maximumTouchDistance * RageTapManager.Constants.maximumTouchDistance){
+            return true;
+        }
+        
+        return false
+    }
+    
+    func tryToFinalizeChain(identifier: ObjectIdentifier){
+        guard let chain = rageTapChainPerUIElement[identifier] else {
+            return
+        }
+        
+        guard let subscriber = subscriber else {
+            return
+        }
+        
+        var tapsWithinTimeout = [TimeInterval]()
+        
+        for tap in chain.actionCommands
+        {
+            tapsWithinTimeout.append(tap.time.timeIntervalSince1970)
+
+            if(tapsWithinTimeout.count == RageTapManager.Constants.minimumRageClicksPerSecond){
+                guard let firstTapTime = tapsWithinTimeout.first?.toInt64Milliseconds else {
+                    break
+                }
+                guard let lastTapTime = tapsWithinTimeout.last?.toInt64Milliseconds else {
+                    break
+                }
+                
+                let timeDiff = lastTapTime - firstTapTime
+                if (timeDiff < RageTapManager.Constants.rageTapWindowTimeout.toInt64Milliseconds)
+                {
+                    // Create rage tap at the last tap of the chain,
+                    // even though we already had matching conditions before that
+                    guard var lastCommand = chain.actionCommands.last else {
+                        break
+                    }
+                    lastCommand.isRage = true
+
+                    subscriber.process(command: lastCommand)
+                    return
+                    
+                }else{
+                    tapsWithinTimeout.remove(at: 0)
+                }
+            }
+        }
+        
+        // Failed to identify chain, send withholded events
+        for tap in chain.actionCommands
+        {
+            subscriber.process(command: tap)
+        }
+    }
+    
+    func addCommandIntoChain(command:RUMAddUserActionCommand, chain: RageTapChain, touch: UITouch){
+        chain.actionCommands.append(command)
+        chain.lastTapPosition = touch.location(in:nil)
+    }
+    
+    func startTimeoutTimerForChain(chain: RageTapChain, identifier:ObjectIdentifier){
+        let workItem = DispatchWorkItem { self.tryToFinalizeChain(identifier: identifier) }
+        chain.chainTimeout = workItem
+        DispatchQueue.global().asyncAfter(deadline: .now() + RageTapManager.Constants.rageTapWindowTimeout, execute: workItem)
```

Comment:
Not really a fan of timer here as it comes with some problems:
* We dispatch the `workItem` to another queue (`.global()`), meaning that `tryToFinalizeChain(identifier:)` will run on another thread. That method relies a lot on the internal state of `RageTapManager` (e.g. mutable `RageTapChain`), so it will eventually crash because the same state is accessed on the main thread in `processActionCommand(command:)`. To make it thread safe, we'd need to add dedicated `queue` for synchronising `RageTapManager`. That would mean introducing at worst 2 more threads to the SDK.
* Timers are painful in tests. Rage taps unit tests would have to await enough time to fulfil the expectation, eventually introducing flakiness.